### PR TITLE
feat(bigquery/storage/managedwriter): decouple connections and writers

### PR DESCRIPTION
**Pull request by shollyman**
This PR is extensive, and represents the cutover to the new abstractions in managedwriter that decouple connection management from writer management.  This PR doesn't expose connection sharing (multiplexing) directly to users at this time, but is a precondition for that feature.

A summary of the changes present in this PR:

### testing changes
As functionality has relocated, so too have many of the tests.  This PR also adds new tests that exercise expanding functionality.

### changes to the internal append flow, and flow control.
The move to a multiple-connection world necessitates changes to the append flow.  appendWithRetry more correctly lives on the writer rather than connection, because a retry may necessitate a change of connection.

Similarly, flow control was handled more coarsely, and that's now been updated. `lockingAppend` is where flow control allocation now happens, and the recv processor is where release happens.  We previously passed in flow controller references as part of the `pendingWrite.markDone()` and that's been removed.

### processRetry
The retryer lives on the writer, but is invoked as part of the receive processor.  There's more refinement needed here, as in a multiplex work we may be using the receive processor for one connection to re-insert writes to another, which may yield undesirable behaviors. 

### revamp send optimizers
This PR revisits send optimization yet again, and switches  behavior so that an internal `pendingWrite` builds and retains a simplex-optimized request by default, and adds the tracking fields necessary to create a fully specified request alongside it.  This allows the dominant case in normal operation to require no extra work on send, which vastly reduces the cpu cycles and amount of memory allocations due to the need for cloning protos in order to perform modifications.

### multiplex e2e
This PR does test multiplex correctness, but none of the work to enable load balancing or pool growth is part of this PR.  The goal here is to mimic existing behavior which expects a 1:1 relationship between writer and it's connection.  Further expansion on multiplexing will be part of subsequent PRs.

### expose GetWriteStream on veneer client
This is small change, but we do now expose the underlying `GetWriteStream` rpc as part of the wrapped client.  It's now much more useful and can return information about resources like default streams.

### updated Close() semantics
The change to having multiple components for servicing requests (ManagedStream, connectionPool, connection, router) represents changes to what `Close()` on a writer means.  For a single-writer pool, we try to shutdown everything when the writer detaches from the pool.

We take a simplistic approach to the `Close()` on the `Client`:  we close the underlying raw client.  Normal error semantics should put the instantiated components into a terminal error state eventually (subject to long-running streams, etc).  A TODO remains to evaluate if we should more aggressively tear down components.

Towards: https://github.com/googleapis/google-cloud-go/issues/7103

### CODE REVIEW

File: bigquery/storage/managedwriter/connection.go [hunk 1]
```diff
@@ -216,21 +285,34 @@ func (co *connection) lockingAppend(pw *pendingWrite) error {
 	var ch chan *pendingWrite
 	var err error
 
-	arc, ch, err = co.getStream(arc, false)
+	// We still need to reconnect if we need to signal a new schema for explicit streams.
+	// Rather than adding more state to the connection, we just look at the request as we
+	// do not allow multiplexing to include explicit streams.
+	forceReconnect := false
+	if !isDefaultStream(pw.writeStreamID) {
```
Comment (alvarowolfx) [inline bigquery/storage/managedwriter/connection.go:292]: seems like in the cases that `isDefaultStream` is being called, what is actually being checked is if the stream supports multiplexing. Maybe change the method name and internally for now it checks if is the default stream.

**Review by alvarowolfx — COMMENTED** [review]
Made some minor comments, but so far it mostly makes sense. Might be nice to have some drawing on how the append request flows between the ManagedStream, Pool, Optimizer and Connections, so it's easier to understand the new flow.

File: bigquery/storage/managedwriter/connection.go (hunk 1 above)
Comment (shollyman) [reply bigquery/storage/managedwriter/connection.go]: renamed to `canMultiplex`.

File: bigquery/storage/managedwriter/client.go [hunk 2]
```diff
@@ -179,6 +202,67 @@ func (c *Client) validateOptions(ctx context.Context, ms *ManagedStream) error {
...
+		}
+		// No existing pool available, create one for the location and add to shared pools.
+		pool, err := c.createPool(ctx, nil, streamFunc, router, true)
+		if err != nil {
+			return nil, err
+		}
+		c.pools[loc] = pool
+		return pool, nil
+	}
+	return c.createPool(ctx, settings, streamFunc, router, false)
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/storage/managedwriter/client.go:226]: resolvePool called during writer instantiation in the client.  Currently (we only have simpleRouter at this point), a pool only can have a single connection associated with it.  Thus, we pair off a pool per writer.  When we introduce a more advanced router capable of maintaining complex connection maps, we can consider always using a pool per region, and having a mix of dedicated and shared connections available to them.

File: bigquery/storage/managedwriter/client.go [hunk 3]
```diff
@@ -70,17 +80,22 @@ func NewClient(ctx context.Context, projectID string, opts ...option.ClientOptio
...
-	}
+
+	// TODO:  Consider if we should actively close pools on client close.
+	// The underlying client closing will cause future calls to the underlying client
+	// to fail terminally, but long-lived calls like an open AppendRows stream may remain
+	// viable for some time.
+	//
+	// If we do want to actively close pools, we need to maintain a list of the singleton pools
+	// as well as the regional pools.
 	c.rawClient.Close()
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/storage/managedwriter/client.go:98]: I think if you want to cleanly shutdown, you close the writers then the client.  Killing the client with live writers is more of a "pull the plug", and the writers will fail as they interact with the raw client.

We can also iterate on this down the line if we want to revisit it.

File: bigquery/storage/managedwriter/appendresult.go [hunk 4]
```diff
@@ -188,43 +189,66 @@ type pendingWrite struct {
...
-					},
-				},
-			},
-		},
-		result: newAppendResult(appends),
+		writer: src,
+		result: newAppendResult(),
 		reqCtx: ctx,
+
+		optimizedRequest: req,
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/storage/managedwriter/appendresult.go]: I switched the field back to a shorter `req` name.

File: bigquery/storage/managedwriter/appendresult.go [hunk 5]
```diff
@@ -188,43 +189,66 @@ type pendingWrite struct {
...
+		descVersion:      curDescVersion,
+		writeStreamID:    writeStreamID,
+		traceID:          traceID,
+	}
+	// Compute the approx size for flow control purposes.
+	pw.reqSize = proto.Size(pw.optimizedRequest) + len(writeStreamID) + len(traceID)
+	if pw.descVersion != nil {
+		pw.reqSize = pw.reqSize + proto.Size(pw.descVersion.descriptorProto)
```
[the remark this answers is not in the document]
Comment (shollyman) [reply bigquery/storage/managedwriter/appendresult.go]: done.

**Review by yirutang — COMMENTED** [review]
With this CL we just have the basic infra ready but real mulitplexing is not enabled?

**Comment by shollyman** [comment]
> With this CL we just have the basic infra ready but real mulitplexing is not enabled?

Correct.  The options are plumbed in, but unexported.  The goal is to migrate to the new abstraction first, then expose the feature publicly with the load aware routing etc.

**Review by alvarowolfx — APPROVED** [review]

**Pull request merged**